openedx/core/djangoapps/programs/utils.py

```diff
@@ -127,6 +128,45 @@ def engaged_programs(self):
 
         return programs
 
+    def _is_course_in_progress(self, now, enrolled_run_modes, course):
+        """Check if course qualifies as in progress as part of the program.
+
+        A course is considered to be in progress if a user is enrolled in a run
+        of the correct mode or a run of the correct mode is still available for enrollment.
+
+        Arguments:
+            now (datetime): datetime for now
+            enrolled_run_modes (dict): Dict showing what mode is enrolled for each run.
+            course (dict): Containing nested course runs.
+
+        Returns:
+            bool, indicating whether the course is in progress.
+        """
+        # Part 1: Check if any of the seats you are enrolled in qualify this course as in progress
+        enrolled_runs = [run for run in course['course_runs'] if run['key'] in self.course_run_ids]
+        # Check if the user is enrolled in the required mode for the run
+        runs_with_required_mode = [
+            run for run in enrolled_runs
+            if run.get('type') == enrolled_run_modes[run['key']]
+        ]
+        if runs_with_required_mode:
+            # Check if the runs you are enrolled in with the right mode are not failed
+            not_failed_runs = [run for run in runs_with_required_mode if run not in self.failed_course_runs]
+            if not_failed_runs:
+                return True
+        # Part 2: Check if any of the seats you are not enrolled in
+        # in the runs you are enrolled in qualify this course as in progress
+        upgrade_deadlines = []
+        for run in enrolled_runs:
+            for seat in run['seats']:
+                if seat['type'] == run.get('type') and run.get('type') != enrolled_run_modes[run['key']]:
+                    upgrade_deadlines.append(seat['upgrade_deadline'])
+
+        course_still_upgradeable = any(
```

Comment:
imo the variable is helpful to understand what is going on, rather than having to understand the any statement